pyquil/quilbase.py

```diff
@@ -1307,76 +1308,196 @@ def nonblocking(self, nonblocking: bool):
+
+    def get_qubits(self, indices: bool = True) -> Union[Set[QubitDesignator], Set[int]]:
+        if indices:
+            return self.get_qubit_indices()
+        return set(self.frame.qubits)
+
+    def get_qubit_indices(self) -> Set[int]:
+        return {qubit.to_fixed() for qubit in super().frame.qubits}
+
+
+class SetPhase(quil_rs.SetPhase, AbstractInstruction):
+    def __new__(cls, frame: Frame, phase: ParameterDesignator):
+        return super().__new__(cls, frame, _convert_to_rs_expression(phase))
 
     def out(self) -> str:
-        return f"SET-PHASE {self.frame} {self.phase}"
+        return str(self)
 
-    def get_qubits(self, indices: bool = True) -> Set[QubitDesignator]:
-        return _get_frame_qubits(self.frame, indices)
+    @property
+    def frame(self):
+        return Frame._from_rs_frame_identifier(super().frame)
+
+    @frame.setter
+    def frame(self, frame: Frame):
+        quil_rs.SetPhase.frame.__set__(self, frame)
+
+    @property
+    def phase(self):
+        return _convert_to_py_parameter(super().phase)
+
+    @phase.setter
+    def phase(self, phase: ParameterDesignator):
+        quil_rs.SetPhase.phase.__set__(self, _convert_to_rs_expression(phase))
+
+    def get_qubits(self, indices: bool = True) -> Union[Set[QubitDesignator], Set[int]]:
+        if indices:
+            return self.get_qubit_indices()
+        return set(self.frame.qubits)
+
+    def get_qubit_indices(self) -> Set[int]:
+        return {qubit.to_fixed() for qubit in super().frame.qubits}
 
 
-class ShiftPhase(AbstractInstruction):
-    def __init__(self, frame: Frame, phase: ParameterDesignator):
-        self.frame = frame
-        self.phase = phase
+class ShiftPhase(quil_rs.ShiftPhase, AbstractInstruction):
+    def __new__(cls, frame: Frame, phase: ParameterDesignator):
+        return super().__new__(cls, frame, _convert_to_rs_expression(phase))
 
     def out(self) -> str:
-        return f"SHIFT-PHASE {self.frame} {self.phase}"
+        return str(self)
 
-    def get_qubits(self, indices: bool = True) -> Set[QubitDesignator]:
-        return _get_frame_qubits(self.frame, indices)
+    @property
+    def frame(self):
+        return Frame._from_rs_frame_identifier(super().frame)
 
+    @frame.setter
+    def frame(self, frame: Frame):
+        quil_rs.ShiftPhase.frame.__set__(self, frame)
+
+    @property
+    def phase(self):
+        return _convert_to_py_parameter(super().phase)
+
+    @phase.setter
+    def phase(self, phase: ParameterDesignator):
+        quil_rs.ShiftPhase.phase.__set__(self, _convert_to_rs_expression(phase))
+
+    def get_qubits(self, indices: bool = True) -> Union[Set[QubitDesignator], Set[int]]:
+        if indices:
+            return self.get_qubit_indices()
+        return set(self.frame.qubits)
+
+    def get_qubit_indices(self) -> Set[int]:
+        return {qubit.to_fixed() for qubit in super().frame.qubits}
+
+
+class SwapPhase(quil_rs.SwapPhases, AbstractInstruction):
```

Comment:
```suggestion
class SwapPhases(quil_rs.SwapPhases, AbstractInstruction):
```

Right? breaking change but actually just a fix. And since it's not supported on hardware or the sim today this is unlikely to break anyone, but it's going to be important to have right in the near future

Reply:
Yup! Good call, renamed it and labeled the commit as a breaking change.